artio-codecs/src/main/java/uk/co/real_logic/artio/fields/DecimalFloat.java

```diff
@@ -189,4 +230,151 @@ public int compareTo(final DecimalFloat other)
             Long.compare(value, other.value) :
             !isPositive ? -1 * scaleComparison : scaleComparison;
     }
+
+    public double toDouble()
+    {
+        if (isNaNValue())
+        {
+            return DOUBLE_NAN_VALUE;
+        }
+        return toDouble(value, scale);
+    }
+
+    public boolean fromDouble(final double doubleValue)
+    {
+        if (Double.isNaN(doubleValue))
+        {
+            value = VALUE_NAN_VALUE;
+            scale = SCALE_NAN_VALUE;
+            return true;
+        }
+        if (!Double.isFinite(doubleValue) ||
+            isOutsideLimits(doubleValue, VALUE_MIN_VAL_AS_DOUBLE, VALUE_MAX_VAL_AS_DOUBLE))
+        {
+            return false;
+        }
+        if (doubleValue == 0.0)
+        {
+            value = 0L;
+            scale = 0;
+            return true;
+        }
+        final boolean isNegative = doubleValue < 0.0;
+        double remainingValue = Math.abs(doubleValue);
+        long newValue = (long)remainingValue;
+        // Have to repeat the limit verification, as the test above on the doubleValue is not exact.
+        // It cuts the worst offenders and allows us to cast to a long,
+        // but it lets through bad values within about 64 of the limits.
+        if (isOutsideLimits(newValue, VALUE_MIN_VAL, VALUE_MAX_VAL))
+        {
+            return false;
+        }
+        int newScale = 0;
+        remainingValue -= newValue;
+        if (remainingValue == 0.0)
+        {
+            setAndNormalise(signedValue(isNegative, newValue), newScale);
+            return true;
+        }
+        while (canValueAcceptMoreDigits(newValue, newScale) &&
+            (newValue == 0L ||
+            !isOutsideLimits(remainingValue, FRACTION_LOWER_THRESHOLD, FRACTION_UPPER_THRESHOLD)))
+        {
+            remainingValue *= 10.0;
+            final double digit = Math.floor(remainingValue);
+            remainingValue -= digit;
+            newValue = newValue * 10L + (long)digit;
+            ++newScale;
+        }
+        if (FRACTION_UPPER_THRESHOLD < remainingValue)
+        {
+            newValue++;
+        }
+        setAndNormalise(signedValue(isNegative, newValue), newScale);
+        return true;
+    }
+
+    public boolean isNaNValue()
+    {
+        return value == VALUE_NAN_VALUE && scale == SCALE_NAN_VALUE;
+    }
+
+    private static DecimalFloat getNaN()
+    {
+        final DecimalFloat nanFloat = new DecimalFloat();
+        nanFloat.value = VALUE_NAN_VALUE;
+        nanFloat.scale = SCALE_NAN_VALUE;
+        return nanFloat;
+    }
+
+    private void setAndNormalise(final long value, final int scale)
+    {
+        this.value = value;
+        this.scale = scale;
+        normalise();
+    }
+
+    private void normalise()
+    {
+        final long valueCopy = value;
```

Comment:
You only use those copies in the ArithmeticException whose method references them. Shouldn't the arithemtic exception throw the new values and you reset the fields to the copy that you started with?

Perhaps an easier way to write this code would be to perform all the operations on local variables and then only set them to the fields at the end if they are inside the limits.

Reply:
That's a fair comment.  I'll change to using local variables.